File: examples/python/bulk_operations/cancel.py
# > Setup

from datetime import datetime, timedelta, timezone

from hatchet_sdk import BulkCancelReplayOpts, Hatchet, RunFilter, V1TaskStatus

hatchet = Hatchet()

workflows = await hatchet.workflows.aio_list()

assert workflows.rows

workflow = workflows.rows[0]


# > List runs
workflow_runs = await hatchet.runs.aio_list(workflow_ids=[workflow.metadata.id])

# > Cancel by run ids
workflow_run_ids = [workflow_run.metadata.id for workflow_run in workflow_runs.rows]

bulk_cancel_by_ids = BulkCancelReplayOpts(ids=workflow_run_ids)

await hatchet.runs.aio_bulk_cancel(bulk_cancel_by_ids)

# > Cancel by filters

bulk_cancel_by_filters = BulkCancelReplayOpts(
    filters=RunFilter(
        since=datetime.today() - timedelta(days=1),
        until=datetime.now(tz=timezone.utc),
        statuses=[V1TaskStatus.RUNNING],
        workflow_ids=[workflow.metadata.id],
        additional_metadata={"key": "value"},
    )
)

await hatchet.runs.aio_bulk_cancel(bulk_cancel_by_filters)

File: examples/typescript/bulk_operations/cancel.ts
import { V1TaskStatus } from '@hatchet-dev/typescript-sdk/clients/rest/generated/data-contracts';
import { hatchet } from '../hatchet-client';

async function main() {
  // > Setup
  const workflows = await hatchet.workflows.list();

  if (!workflows.rows?.length) {
    throw new Error('no workflows found');
  }

  const [workflow] = workflows.rows;

  // > List runs
  const workflowRuns = await hatchet.runs.list({
    workflowNames: [workflow.name],
  });

  // > Cancel by run ids
  const runIds = workflowRuns.rows?.map((run) => run.metadata.id) ?? [];

  // to replay runs by their ids, use `hatchet.runs.replay` instead
  await hatchet.runs.cancel({ ids: runIds });

  // > Cancel by filters
  // to replay runs matching filters, use `hatchet.runs.replay` instead
  await hatchet.runs.cancel({
    filters: {
      since: new Date(Date.now() - 24 * 60 * 60 * 1000),
      until: new Date(),
      statuses: [V1TaskStatus.RUNNING],
      workflowNames: [workflow.name],
      additionalMetadata: { key: 'value' },
    },
  });
}

if (require.main === module) {
  main()
    .catch(console.error)
    .finally(() => process.exit(0));
}
